File: src/main/java/core/procedures/math/LCM.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigRatio;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

import static core.procedures.math.GCD.gcd;

public final class LCM extends AFn {

  private static final Abs ABS = new Abs();

  public LCM() {
    super(new FnArgsBuilder().rest(BigRatio.class).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "lcm";
  }

  @Override
  public Number apply(Object... args) {
    if (args.length == 0) {
      return 1L;
    }
    if (args.length == 1) {
      return ABS.apply1(args[0]);
    }
    Number result = (Number) args[0];
    for (int i = 1; i < args.length; i++) {
      result = lcm(result, (Number) args[i]);
    }
    return result;
  }

  private static long lcm(Long a, Long b) {
    if ((a.intValue() == 0) && (b.intValue() == 0)) {
      return 0L;
    }
    return (a / gcd(a, b)) * b;
  }

  private static Double lcm(Double a, Double b) {
    if ((a.intValue() == 0) && (b.intValue() == 0)) {
      return 0d;
    }
    return (a / gcd(a, b).doubleValue()) * b;
  }

  static BigInteger lcm(BigInteger first, BigInteger second) {
    return first.multiply(second.divide(gcd(first, second)));
  }

  private BigRatio lcm(BigRatio first, BigRatio second) {
    return BigRatio.valueOf(lcm(first.getNumerator(), second.getNumerator()),
                           gcd(first.getDenominator(), second.getDenominator()));
  }

  private Number lcm(BigDecimal a, BigDecimal b) {
    if ((BigDecimal.ZERO.compareTo(a) == 0) && (BigDecimal.ZERO.compareTo(b) == 0)) {
      return BigDecimal.ZERO;
    }
    int scale = Math.max(a.scale(), b.scale());
    if (scale == 0) {
      return new BigDecimal(lcm(a.toBigInteger(), b.toBigInteger()));
    } else {
      return ToInexact.toInexact(lcm((BigDecimal)ToExact.toExact(a), (BigDecimal) ToExact.toExact(b)));
    }
  }

  private Number lcm(Number first, Number second) {
    Number f = Utils.INSTANCE.upcast(first);
    Number s = Utils.INSTANCE.upcast(second);
    if ((f instanceof Long) && (s instanceof Long)) {
      return lcm((Long)first, (Long)second);
    }
    if ((first instanceof BigRatio) && (second instanceof BigRatio)) {
      return lcm((BigRatio) first, (BigRatio)second);
    }
    if (first instanceof BigRatio) {
      return lcm(((BigRatio) first).toBigDecimal(), Utils.INSTANCE.toBigDecimal(second));
    }
    if (second instanceof BigRatio) {
      return lcm(Utils.INSTANCE.toBigDecimal(first), ((BigRatio) second).toBigDecimal());
    }
    if ((first instanceof BigDecimal) || (second instanceof BigDecimal)) {
      return lcm(Utils.INSTANCE.toBigDecimal(first), Utils.INSTANCE.toBigDecimal(second));
    }
    if (first instanceof BigInteger || second instanceof BigInteger) {
      return lcm(Utils.INSTANCE.toBigInteger(first), Utils.INSTANCE.toBigInteger(second));
    }
    return lcm(first.doubleValue(), second.doubleValue());
  }
}

File: src/main/java/core/procedures/math/LCM.kt
package core.procedures.math

import core.procedures.AFn
import core.scm.Ratio
import core.scm.Type
import core.utils.Utils

import java.math.BigDecimal
import java.math.BigInteger
import kotlin.math.absoluteValue

class LCM : AFn<Any?, Number>(name = "lcm", isPure = true, restArgsType = Type.Rational::class.java) {

    private val toInexact = ToInexact()
    private val toExact   = ToExact()
    private val abs       = Abs()

    override operator fun invoke(args: Array<out Any?>): Number = when (args.size) {
        0    -> 1L
        1    -> abs(args[0]!! as Number)
        else -> args.fold(args[0]!! as Number) { r, n -> lcm(r, n!! as Number) }
    }

    private fun lcm(first: Ratio, second: Ratio) = Ratio.valueOf(lcm(first.numerator, second.numerator),
                                                                          GCD.gcd(first.denominator, second.denominator))

    fun lcm(first: BigInteger, second: BigInteger): BigInteger {
        val f = first.abs()
        val s = second.abs()
        return f * s.divide(GCD.gcd(f, s))
    }

    private fun lcm(a: BigDecimal, b: BigDecimal): Number = when {
        a.signum() == 0 && b.signum() == 0 -> BigDecimal.ZERO
        maxOf(a.scale(), b.scale()) == 0   -> lcm(a.toBigInteger(), b.toBigInteger()).toBigDecimal()
        else -> toInexact(lcm(toExact(a) as BigDecimal, toExact(b) as BigDecimal))
    }

    private fun lcm(a: Long, b: Long) = when {
        a == 0L && b == 0L -> 0L
        else -> a / GCD.gcd(a, b) * b
    }

    private fun lcm(a: Double, b: Double) = when {
        a.toInt() == 0 && b.toInt() == 0 -> 0.0
        else -> a / GCD.gcd(a, b).toDouble() * b
    }

    private fun lcm(first: Number, second: Number): Number {
        val (f, s) = Utils.upcast(first, second)
        return when {
            f is Double     && s is Double     -> lcm(f.absoluteValue, s.absoluteValue)
            f is Float      && s is Float      -> lcm(f.toDouble().absoluteValue, s.toDouble().absoluteValue)
            f is Ratio   && s is Ratio   -> lcm(f, s)
            f is BigDecimal && s is BigDecimal -> lcm(f, s)
            f is BigInteger && s is BigInteger -> lcm(f, s)
            else                               -> lcm(f.toLong().absoluteValue, s.toLong().absoluteValue)
        }
    }
}
